docs/CollectionBestPractices.md

```diff
@@ -0,0 +1,360 @@
+> - If exposing a collection is necessary, consider whether it might be better to expose a more optimal read-only
+>   collection. Instead of `IReadOnlyList<T>`, consider [`ImmutableArray<T>`](#immutablearrayt).
+> - There aren’t many other options when an API calls for exposing an `IReadOnlyDictionary<TKey, TValue>`. In these
+>   cases, consider whether it might be better to just avoid exposing the collection altogether and provide APIs that
+>   access it. Or, in some cases, it might be necessary to create entirely new collection types. (This is why Razor
+>   `TagHelperDescriptors` expose a `MetadataCollection`.)
+
+> [!WARNING]
+> **Be mindful of ToArray()**
+>
+> Calling `ToArray()` on a collection will create a new array and copy content from the collection into it. So, when
+> the exact capacity is known up front, it is an anti-pattern to create a `List<T>` withwout that capacity, fill it
+> with items and then call `ToArray()` at the end. This results in extra allocations that could be avoided by creating
+> an array and filling it.
+
+# Immutable Collections
+- The .NET immutable collections are provided by the System.Collections.Immutable NuGet package, which provides
+  implementations for .NET, .NET Framework, and .NET Standard 2.0.
+- The collections in the System.Collections.Immutable namespace have a very specific purpose.
+- They are intended to be *persistent* data structures; that is, a data structure that always preserves the previous
+  version of itself when it is modified. Such data structures are effectively immutable, but it might have been better
+  for the namespace to have been called, System.Collections.Persistent.
+  - The term “persistent data structure” was introduced by the 1986 paper,
+    “Making Data Structures Persistent” ([PDF](https://www.cs.cmu.edu/~sleator/papers/making-data-structures-persistent.pdf)).
+  - A highly influential book in the area of persistent data structures is “Purely Functional Data Structures” (1999)
+    by Chris Okasaki ([Amazon](https://www.amazon.com/Purely-Functional-Data-Structures-Okasaki/dp/0521663504)).
+    Okasaki’s original dissertation is available from CMU’s website ([PDF](https://www.cs.cmu.edu/~rwh/students/okasaki.pdf)).
+- Because of their persistency, nearly all of the immutable collections have very different implementations than their
+  imperative counterparts. For example, `List<T>` is implemented using an array, while `ImmutableList<T>` is implemented
+  using a binary tree.
+- Mutating methods on an immutable collection perform “non-destructive mutation”. Instead, of mutating the underlying
+  object, a mutating method like `Add` produces a new instance of the immutable collection. This is similar to how the
+  `String.Replace(...)` API is used.
+- The difference in implementation affects the asymptotic complexity of many standard operations. For example, indexing
+  into a `List<T>` is O(1) but indexing into an `ImmutableList<T>` is O(log n).
```

Comment:
Consider adding the perf difference between `Dictionary<TKey, TValue>` and `ImmutableDictionary<TKey, TValue>`. Seen that trip up a number of people (myself included)

Reply:
Good idea! I'll also use a callout for this to draw attention to it.